glue_vispy_viewers/volume/volume_toolbar.py

```diff
@@ -8,19 +8,103 @@
 from glue.utils.geometry import points_inside_poly
 from ..utils import as_matrix_transform
+from matplotlib import path
+from glue.core.roi import RectangularROI, CircularROI, PolygonalROI
+from astrodendro import Dendrogram
+from ..extern.vispy import scene
 
 
 class VolumeSelectionToolbar(VispyDataViewerToolbar):
 
     def __init__(self, vispy_widget=None, parent=None):
         super(VolumeSelectionToolbar, self).__init__(vispy_widget=vispy_widget, parent=parent)
+        # for getting transposed vol_data shape and getting pos array
         self.trans_ones_data = None
+        # add some markers
+        self.markers = scene.visuals.Markers(parent=self._vispy_widget.view.scene)
+        self.ray_line = scene.visuals.Line(color='green', width=5,
+                                           parent=self._vispy_widget.view.scene)
+
+        self.visual_tr = None
+        self.visible_data = None
+        self.visual = None
+        self.vol_data = None
+        self.max_value_pos = None
 
     def on_mouse_press(self, event):
-        self.selection_origin = event.pos
-        if self.mode is 'point':
-            # TODO: add dendrogram selection here
-            pass
+        """
+        Assign mouse position and do point selection.
+        :param event:
+        """
+        if self.mode:
+            # do the initiation here
+            self.selection_origin = event.pos
+            self.visible_data, self.visual = self.get_visible_data()
+
+            # for dendrogram cube the two attributes are intensity and structure, no ['PRIMARY']
+            data_array = self.visible_data[0]['intensity']
+
+            self.trans_ones_data = np.transpose(np.ones(data_array.shape))
+
+            self.vol_data = data_array # current_layer
+            self.visual_tr = self._vispy_widget.limit_transforms[self.visual]
+
+        if self.mode is 'dendrogram':
+
+            # get start and end point of ray line
+            pos = self.get_ray_line()
+            max_value_pos, max_value = self.get_inter_value(pos)
+            self.max_value_pos = max_value_pos
+
+            # change the color of max value marker
+            if max_value:
+                self.markers.set_data(pos=np.array(max_value_pos),
+                                      face_color='yellow')
+                status_text = 'pos '+str(max_value_pos[0]) \
+                              + ' value '+str(max_value)
+                # TODO: AttributeError: 'VispyVolumeViewer' object has no attribute 'show_status'
+                # self._vispy_data_viewer.show_status(status_text)
+
+                # TODO: vispy_viewer doesn't have client, how to get dendrogram data?
+                print('visible_data, client.data', self.visible_data, self._vispy_widget.client.data)
+                for each_data in self._vispy_widget.client.data:
+                    if each_data.label == 'Dendrogram':
+                        d = each_data
+
```

Comment:
@astrofrog # TODO: vispy_widget doesn't have a client, how to get dendrogram data here?